### ISSUE TITLE
CTCLoss backward is inconsistent with its forward derivative with respect to the documented log_probs input.

### ISSUE THREAD

**Issue by beanduan22** [pytorch/pytorch#181534]
### 🐛 Describe the bug

```py
import torch
import torch.nn as nn
import torch.nn.functional as F

torch.manual_seed(123)

T, N, C = 3, 1, 3  # time, batch, classes incl. blank
log_probs = F.log_softmax(torch.randn(T, N, C), dim=-1).double()
log_probs.requires_grad_(True)

targets = torch.tensor([[1]])
input_lens = torch.tensor([T])
target_lens = torch.tensor([1])

loss_fn = nn.CTCLoss(blank=0)

def f(x):
    return loss_fn(x, targets, input_lens, target_lens)

print(torch.autograd.gradcheck(f, (log_probs,), raise_exception=True))
```

Environment:
- PyTorch: 2.11.0+cu128
- Python: 3.13.5

The following minimal example reproduces the issue. `gradcheck` fails with a clear Jacobian mismatch for the input `log_probs`.

The numerical Jacobian contains exact zeros for class 2 at every time step:

  numerical:
  tensor([[-0.4727],
          [-0.5273],
          [ 0.0000],
          [-0.5624],
          [-0.4376],
          [ 0.0000],
          [-0.4784],
          [-0.5216],
          [ 0.0000]], dtype=torch.float64)

However, the analytical gradient returned by CTCLoss backward gives non-zero values at the same positions:

  analytical:
  tensor([[-0.1430],
          [-0.1116],
          [ 0.2547],
          [-0.2236],
          [-0.3075],
          [ 0.5311],
          [-0.3088],
          [-0.3109],
          [ 0.6198]], dtype=torch.float64)

Since class 2 is neither the blank class nor the target label in this example, the true derivative with respect to `log_probs[..., 2]` should be zero. The numerical Jacobian confirms this, but the analytical backward returns non-zero gradients.

This suggests that `CTCLoss.backward()` is inconsistent with the derivative of `CTCLoss.forward()` with respect to the documented `log_probs` input.

### Versions

PyTorch: 2.11.0+cu128
Python: 3.13.5

cc @ezyang @gchanan @kadeng @msaroufim @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by drisspg** [comment]
reproduces on nightly for me

**Comment by jashshah999** [comment]
I investigated this and I believe the backward is actually correct *for the intended use case*, but the documentation is misleading.

**What's happening:** The CTC backward computes `grad[t,c] = prob[t,c] - posterior[t,c]`, which is the gradient through a softmax — i.e., `d(loss)/d(logit)` where `prob = softmax(logit)`. For classes not in any CTC path, `posterior = 0`, so the gradient equals `exp(log_prob[t,c])`.

This is consistent with the original Graves (2006) CTC formulation which differentiates through the softmax layer. The intended usage is:

```python
loss = ctc_loss(log_softmax(logits), ...)  # correct: gradcheck passes
loss.backward()  # gives correct d(loss)/d(logits) via chain rule
```

And indeed `gradcheck` passes for this:

```python
def f(logits):
    return F.ctc_loss(F.log_softmax(logits, dim=-1), targets, input_lens, target_lens)

torch.autograd.gradcheck(f, (logits,))  # True
```

But fails when called directly on log-probabilities because the backward already absorbs the softmax Jacobian.

**Evidence:** The gradients sum to exactly zero per timestep (the hallmark of softmax-absorbed gradients), and for non-target classes, `grad[t,c]` exactly equals `exp(log_prob[t,c])`.

This is arguably a **documentation issue** rather than a code bug — the docs say the input is "log-probabilities" but don't mention that backward treats them as log-softmax outputs (absorbing the Jacobian). The behavior matches `nn.CrossEntropyLoss` which similarly absorbs the softmax into the backward.

That said, an argument can be made that the backward should compute the mathematically correct `d(loss)/d(log_prob)` and let autograd handle the softmax chain rule — this would make the API more composable. The tradeoff is backward compatibility.

**Comment by beanduan22** [comment]
Thanks for taking a look — I think your explanation matches what I’m seeing.

I agree the current backward seems to be correct for the common usage.

But the confusing part is that the API/documentation says the input is `log_probs`. If a user treats `log_probs` as the actual differentiable input, then the backward doesn’t match the forward derivative, and `gradcheck` fails. I think the docs should make this behavior explicit.

**Comment by MSK-005** [comment]
Will create a PR to fix the documentation.
